### DISCUSSION TITLE
Relationships with viewonly=true and raise="lazy" are not preserved during a merge with load=false

### DISCUSSION THREAD

**Discussion by twiddli** [sqlalchemy/sqlalchemy#8861] [Usage Questions] [answered]
Hello!
I'm not quite sure whether this is a bug or not but I have several relationships defined with viewonly=True that are causing issues when I try to access them after moving objects between different sessions.
The issue is similar to #3488, which suggests this should work. I especially can't find anything in the docs about viewonly=True not being able to work with merge. Here's a minimal example:

```
from sqlalchemy import *
from sqlalchemy.orm import *

Base = declarative_base()

class C(Base):
    __tablename__ = 'c'
    id = Column(Integer, primary_key=True)
    a_id = Column(Integer, ForeignKey('a.id'), nullable=False)

class A(Base):
    __tablename__ = 'a'
    id = Column(Integer, primary_key=True)
    c = relationship(C, primaryjoin=lambda: A.id==C.a_id, lazy="raise", viewonly=True)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

a1 = A()
c = C()

s = Session(e)
s.add(a1)
s.commit()
c.a_id = a1.id
s.add(c)
s.commit()

a1 = s.query(A).options(joinedload(A.c)).first()

assert a1.c # works

s.close()

assert inspect(a1).detached

assert a1.c # works

s2 = Session(e)
m_a1 = s2.merge(a1, load=False)

print("---------------")
print(m_a1.c) # fails with sqlalchemy.exc.InvalidRequestError: 'A.c' is not available due to lazy='raise'
```

If I remove the `lazy="raise` parameter, it would resolve to None instead, which is also undesirable.
Viewonly relationships are really useful for a variety of reasons, so I would rather not resort to not using them. Is there any way to move an object to a different session while fully preserving all loaded attributes?

> Off-topic but It would be nice to have a fast merge function that associates an unmodified object as-is, as if it were loaded with that session in the first place.
> A common use-case pattern I'm experiencing is having a function return a detached object, which you want to do further work on and then commit the changes to the database.

**Comment by zzzeek** [comment, marked as answer]
that looks completely like a bug.  very strange, as the purpose of the merge/load=False is exactly as you are using it and the viewonly shouldnt change loader status, i will open that now.

> Off-topic but It would be nice to have a fast merge function that associates an unmodified object as-is, as if it were loaded with that session in the first place.

if the object you have is detached, you can!  just add() it to whatever session you'd like.  you can do that right here:

```py
s2 = Session(e)
s2.add(a1)

```

caveats!   if there is already an object with that identity in that Session, it will raise.  secondly, you can't add "a1" to more than one Session at a time.

**Reply by zzzeek** [reply]
#8862

**Reply by twiddli** [reply]
Thank you so much for the response!
